data-protocols/ids/ids-transform-v1/src/main/java/org/eclipse/dataspaceconnector/ids/transform/IdsLogicalConstraintToConstraintTransformer.java

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright (c) 2021 Daimler TSS GmbH
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Daimler TSS GmbH - Initial Implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.ids.transform;
+
+import de.fraunhofer.iais.eis.LogicalConstraint;
+import org.eclipse.dataspaceconnector.common.collection.CollectionUtil;
+import org.eclipse.dataspaceconnector.ids.spi.transform.IdsTypeTransformer;
+import org.eclipse.dataspaceconnector.ids.spi.transform.TransformerContext;
+import org.eclipse.dataspaceconnector.policy.model.AndConstraint;
+import org.eclipse.dataspaceconnector.policy.model.Constraint;
+import org.eclipse.dataspaceconnector.policy.model.MultiplicityConstraint;
+import org.eclipse.dataspaceconnector.policy.model.OrConstraint;
+import org.eclipse.dataspaceconnector.policy.model.XoneConstraint;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+public class IdsLogicalConstraintToConstraintTransformer implements IdsTypeTransformer<LogicalConstraint, Constraint> {
+
+    @Override
+    public Class<LogicalConstraint> getInputType() {
+        return LogicalConstraint.class;
+    }
+
+    @Override
+    public Class<Constraint> getOutputType() {
+        return Constraint.class;
+    }
+
+    @Override
+    public @Nullable Constraint transform(LogicalConstraint constraint, @NotNull TransformerContext context) {
+        Objects.requireNonNull(context);
+        if (constraint == null) {
+            return null;
+        }
+        int constraintTypes = 0;
+        if (CollectionUtil.isNotEmpty(constraint.getAnd())) {
+            constraintTypes++;
+        }
+        if (CollectionUtil.isNotEmpty(constraint.getOr())) {
+            constraintTypes++;
+        }
+        if (CollectionUtil.isNotEmpty(constraint.getXone())) {
+            constraintTypes++;
+        }
+        if (constraintTypes == 0) {
+            context.reportProblem("Logical constraint must specify at least AND, OR, or XONE constraints");
+            return null;
+        } else if (constraintTypes == 1) {
+            return transformSingleConstraintType(constraint, context);
+        } else {
+            return transformMultipleConstraintTypes(constraint, context);
+        }
```

Comment:
Code here seems a little bit complex, but don't have a idea on how to simplify right now. Let's improve this within our refactoring week.

Reply:
Sure. Unfortunately, however, I think this complexity is a result of the InfoModel not using an inheritance hierarchy for Constraints. That choice also leads to other issues, including the difficulty of supporting nested constraints through recursion in IDS. My initial inclination is the InforModel itself will need to change for this to get simpler.

There is also #338 which IMO is a critical blocker.